metal/resource_metal_port.go

```diff
@@ -0,0 +1,206 @@
+			},
+			"network_type": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "One of layer2-bonded, layer2-individual, layer3, hybrid and hybrid-bonded",
+			},
+			"disbond_supported": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Flag indicating whether the port can be removed from a bond",
+			},
+			"bond_name": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Name of the bond port",
+			},
+			"bond_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "UUID of the bond port",
+			},
+			"type": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Port type",
+			},
+			"mac": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "MAC address of the port",
+			},
+		},
+	}
+}
+
+func resourceMetalPortUpdate(d *schema.ResourceData, meta interface{}) error {
+	cpr, _, err := getClientPortResource(d, meta)
+	if err != nil {
+		return friendlyError(err)
+	}
+
+	for _, f := range [](func(*ClientPortResource) error){
+		portSanityChecks,
+		batchVlans(true),
+		makeDisbond,
+		convertToL2,
+		makeBond,
+		convertToL3,
+		batchVlans(false),
+	} {
+		if err := f(cpr); err != nil {
+			return friendlyError(err)
+		}
+	}
+
+	return resourceMetalPortRead(d, meta)
+}
+
+func resourceMetalPortRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+	port, err := getPortByResourceData(d, client)
+
+	if err != nil {
+		if isNotFound(err) || isForbidden(err) {
+			log.Printf("[WARN] Port (%s) not accessible, removing from state", d.Id())
+			d.SetId("")
+
+			return nil
+		}
+		return err
+	}
+	m := map[string]interface{}{
+		"port_id":           port.ID,
+		"type":              port.Type,
+		"name":              port.Name,
+		"network_type":      port.NetworkType,
+		"mac":               port.Data.MAC,
+		"bonded":            port.Data.Bonded,
+		"disbond_supported": port.DisbondOperationSupported,
+	}
+	l2 := contains(l2Types, port.NetworkType)
+	l3 := contains(l3Types, port.NetworkType)
+
+	if l2 {
+		m["layer2"] = true
+	}
+	if l3 {
+		m["layer2"] = false
+	}
+
+	if port.NativeVirtualNetwork != nil {
+		m["native_vlan_id"] = port.NativeVirtualNetwork.ID
+	}
+
+	vlans := []string{}
+	for _, n := range port.AttachedVirtualNetworks {
+		vlans = append(vlans, n.ID)
+	}
+	m["vlan_ids"] = vlans
+
+	if port.Bond != nil {
+		m["bond_id"] = port.Bond.ID
+		m["bond_name"] = port.Bond.Name
+	}
+
+	d.SetId(port.ID)
+	return setMap(d, m)
+}
+
+func resourceMetalPortDelete(d *schema.ResourceData, meta interface{}) error {
+	resetRaw, resetOk := d.GetOk("reset_on_delete")
+	if resetOk && resetRaw.(bool) {
+		cpr, resp, err := getClientPortResource(d, meta)
+		if ignoreResponseErrors(httpForbidden, httpNotFound)(resp, err) != nil {
+			return err
+		}
+
+		for _, f := range [](func(*ClientPortResource) error){
+			removeNativeVlan,
```

Comment:
you can't use `removeNativeVlan`, `makeBond` and `convertToL3` here. Those functions will make sure that the different aspects of the port configuration (native VLAN, bond, L3) are set according to the resource conf. That's why the `schema.ResourceData` is part of the CPR struct. If you call e.g. `convertToL3` to a resource

```
resource "metal_port" "bond0" {
  port_id = local.bond0_id
  layer2 = true
  bonded = true
  reset_on_delete = true
```
.. it will not convert the port to L3 because in the resource conf, layer2 is true.

What you want to achieve here is to put the port to somehow default state: no vlans, no native vlan, L3, bonded. You don't need the Resource struct for that, you shouldn't look at the current configuration. 

The function should go like

```golang
func portToDefault(c *packngo.Client, portID string) error {
	getOpts := &packngo.GetOptions{Includes: []string{
		"native_virtual_network",
		"virtual_networks",
	}}
	port, _, err := c.Ports.Get(portID, getOpts)
	if err != nil {
		return err
	}

	if port.NativeVirtualNetwork != nil {
		// remove native VLAN with the new bulk API
	}

	if len(port.AttachedVirtualNetworks) > 0 {
		// remove VLANs with the new bulk API
	}

	// create bond
	if !port.Data.Bonded {
		port, _, err = c.DevicePorts.Bond(port, false)
		if err != nil {
			return err
		}
	}
	if contains(l2Types, port.NetworkType) {
		ips := []packngo.AddressRequest{
			{AddressFamily: 4, Public: true},
			{AddressFamily: 4, Public: false},
			{AddressFamily: 6, Public: true},
		}
		_, _, err := c.Ports.ConvertToLayerThree(port.ID, ips)
		if err != nil {
			return err
		}
	}
	return nil
}
```

Reply:
@displague ^

Reply:
You can then have the delete function for metal_port as:

```golang
func resourceMetalPortDelete(d *schema.ResourceData, meta interface{}) error {
        client := meta.(*packngo.Client)
	resetRaw, resetOk := d.GetOk("reset_on_delete")
	if resetOk && resetRaw.(bool) {
		return portToDefault(client, d.Id())
        }
        return nil
}
```

Reply:
One of the concerns I have is in how `reset_on_delete` should behave on `eth0` vs `eth1` vs `bond0` vs `bond1`. We are taking on some authority about how the Equinix Metal API is expected to behave without the Equinix Metal API expressly stating that behavior, or offering an API endpoint to exact the defaults.

I know of, for example, an operating system that is deployed with Layer2-Individual as the default port configuration.

Perhaps this is just a documentation concern.

> `reset_on_delete` will reset the port to the most common default values when this resource is deleted.  This setting may impact other ports on the same bond.  For eth0, eth1, and bond0 ports, the port will be bonded (with both eth0 and eth1) in Layer3 mode, potentially issuing new IP addresses, with no VLANs attached.  For eth2, eth3, and bond1 ports, the port will be converted to Layer2-Bonded.

Reply:
@displague hmm, I see, the defaults might be different. Well we could document that reset_on_default will mean that the bond will go to bonded L3, and users shouldn't be using reset_on_delete with the device type which defaults to L2-bonded.

Reply:
All but the L2Bonded tests passed. I'll pick that up next:

```
    resource_metal_port_test.go:128: Step 4/5 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # metal_port.bond0 will be updated in-place
          ~ resource "metal_port" "bond0" {
                bonded            = true
                disbond_supported = true
                id                = "c20dbb55-5d59-4de8-9571-be11c7968ac9"
              - layer2            = true -> null
                name              = "bond0"
                network_type      = "layer2-bonded"
                port_id           = "c20dbb55-5d59-4de8-9571-be11c7968ac9"
                type              = "NetworkBondPort"
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
```

Reply:
@displague I think it's because the reset on delete doesn't really happen. The port is not put back to L3 (You can check the POST calls in the standard output when run with TF_LOG=DEBUG and TF_LOG_PATH=logfile.txt). It's because of the reasons I wrote here:
https://github.com/equinix/terraform-provider-metal/pull/116/files#r686799824

Reply:
The tests are passing. I'm wondering if bonded and layer2 should be optional+computed.  The error we were getting before about `layer3 -> null` is because layer3 is only set to true or false depending on if `l2` or l3` is detected, but it is possible that neither is detected.